fgpyo/sam/__init__.py

```diff
@@ -606,6 +607,110 @@ def __getitem__(
         return self.elements[index]
 
 
+@enum.unique
+class PairOrientation(enum.Enum):
+    """Enumerations of read pair orientations."""
+
+    FR = "FR"
+    """A pair orientation for forward-reverse reads ("innie")."""
+
+    RF = "RF"
+    """A pair orientation for reverse-forward reads ("outie")."""
+
+    TANDEM = "TANDEM"
+    """A pair orientation for tandem (forward-forward or reverse-reverse) reads."""
+
+    @classmethod
+    def build(
+        cls, r1: AlignedSegment, r2: Optional[AlignedSegment] = None
+    ) -> Optional["PairOrientation"]:
+        """Returns the pair orientation if both reads are mapped to the same reference sequence.
+
+        Args:
+            r1: The first read in the template.
+            r2: The second read in the template. If undefined, mate data set upon R1 will be used.
+
+        See:
+            [`htsjdk.samtools.SamPairUtil.getPairOrientation()`](https://github.com/samtools/htsjdk/blob/c31bc92c24bc4e9552b2a913e52286edf8f8ab96/src/main/java/htsjdk/samtools/SamPairUtil.java#L71-L102)
+        """
+
+        if r2 is None:
+            r2_is_unmapped = r1.mate_is_unmapped
+            r2_reference_id = r1.next_reference_id
+        else:
+            r2_is_unmapped = r2.is_unmapped
+            r2_reference_id = r2.reference_id
+
+        if r1.is_unmapped or r2_is_unmapped or r1.reference_id != r2_reference_id:
+            return None
+
+        if r2 is None:
+            if not r1.has_tag("MC"):
+                raise ValueError('Cannot determine proper pair status without R2\'s cigar ("MC")!')
+            r2_cigar = Cigar.from_cigarstring(str(r1.get_tag("MC")))
+            r2_is_forward = r1.mate_is_forward
+            r2_reference_start = r1.next_reference_start
+            r2_reference_end = r1.next_reference_start + r2_cigar.length_on_target()
+        else:
+            r2_is_forward = r2.is_forward
+            r2_reference_start = r2.reference_start
+            r2_reference_end = r2.reference_end
+
+        if r1.is_forward is r2_is_forward:
+            return PairOrientation.TANDEM
+        elif r1.is_forward and r1.reference_start < r2_reference_end:
+            return PairOrientation.FR
+        elif r1.is_reverse and r2_reference_start < r1.reference_end:
+            return PairOrientation.FR
+        else:
+            return PairOrientation.RF
+
+
+DefaultProperlyPairedOrientations = {PairOrientation.FR}
+"""The default orientations for properly paired reads."""
+
+
+def is_proper_pair(
+    r1: AlignedSegment,
+    r2: Optional[AlignedSegment] = None,
+    max_insert_size: int = 1000,
+    orientations: set[PairOrientation] = DefaultProperlyPairedOrientations,
```

Comment:
How would you feel about making this `Iterable` and using a `tuple` for the default?

Reply:
I like making the arg `Iterable` as it it is more generic. But a `set` default value seems to capture the intent more than a non-set (why would you allow duplicates?).

What I think I need to do is use a `Container` type to support a set default value, but other types of containers too:

https://docs.python.org/3/library/collections.abc.html#collections.abc.Container

Reply:
https://github.com/fulcrumgenomics/fgpyo/pull/201/commits/4e64644fc0176bc93bfa32e4ed6f7a9e61118b06